File: providers/multiplexer/ml/EverCrypt_Bytes.ml
open Ctypes
open PosixTypes
open Foreign

(* This module implements EverCrypt.Bytes, using fstarlib's `type bytes =
 * string`. *)

(* See Santiago's comment in PKI.ml *)
let () =
  if Sys.os_type = "Win32" then
    ignore (Dl.dlopen ~filename:"libevercrypt.so" ~flags:[])

let x25519_ =
  foreign "EverCrypt_x25519" (
    ptr char @->
    string @->
    string @->
    returning void)

let x25519 secret base =
  let dst = allocate_n char 32 in
  x25519_ dst secret base;
  string_from_ptr dst 32


type cipher_tag = {
  cipher: string;
  tag:    string
}

let chacha20_poly1305_encrypt_ =
  foreign "EverCrypt_chacha20_poly1305_encrypt" (
    ptr char @->
    ptr char @->
    string @->
    uint32_t @->
    string @->
    uint32_t @->
    string @->
    string @->
    returning uint32_t)

let chacha20_poly1305_encrypt m aad k n =
  let m_len   = String.length m in
  let aad_len = String.length aad in
  let cipher  = allocate_n char m_len in
  let tag     = allocate_n char 16 in
  let _ = chacha20_poly1305_encrypt_ cipher tag
            m   (Unsigned.UInt32.of_int m_len)
            aad (Unsigned.UInt32.of_int aad_len) k n in
  { cipher = string_from_ptr cipher m_len;
    tag    = string_from_ptr tag 16 }

type maybe_plaintext =
| Error
| Correct of string

let chacha20_poly1305_decrypt_ =
  foreign "EverCrypt_chacha20_poly1305_decrypt" (
    ptr char @->
    string @->
    uint32_t @->
    string @->
    string @->
    uint32_t @->
    string @->
    string @->
    returning uint32_t)

let chacha20_poly1305_decrypt c tag aad k n =
  let m_len   = String.length c in
  let aad_len = String.length aad in
  let m       = allocate_n char m_len in
  let res = chacha20_poly1305_decrypt_ m c
            (Unsigned.UInt32.of_int m_len) tag aad
            (Unsigned.UInt32.of_int aad_len) k n in
  if Unsigned.UInt32.to_int res = 0 then
    Correct (string_from_ptr m m_len)
  else Error

File: providers/evercrypt/c/evercrypt_bytes.c
#include "kremlib.h"
#include "EverCrypt_Bytes.h"
#include "EverCrypt.h"

FStar_Bytes_bytes EverCrypt_Bytes_x25519(FStar_Bytes_bytes secret, FStar_Bytes_bytes base) {
  FStar_Bytes_bytes out = {
    .length = 32,
    .data = KRML_HOST_CALLOC(32, 1)
  };
  EverCrypt_x25519((uint8_t *) out.data, (uint8_t *) secret.data,  (uint8_t *) base.data);
  return out;
}

EverCrypt_Bytes_cipher_tag
EverCrypt_Bytes_chacha20_poly1305_encrypt(FStar_Bytes_bytes m,
                                          FStar_Bytes_bytes aad,
                                          FStar_Bytes_bytes k,
                                          FStar_Bytes_bytes n) {
  FStar_Bytes_bytes cipher = {
    .length = m.length,
    .data = KRML_HOST_CALLOC(m.length, 1)
  };
  FStar_Bytes_bytes tag = {
    .length = 16,
    .data = KRML_HOST_CALLOC(16, 1)
  };
  EverCrypt_Bytes_cipher_tag out = {
    .cipher = cipher,
    .tag = tag
  };
  EverCrypt_chacha20_poly1305_encrypt((uint8_t *) k.data,
                                      (uint8_t *) n.data,
                                      (uint8_t *) aad.data,
                                      aad.length,
                                      (uint8_t *) m.data,
                                      m.length,
                                      (uint8_t *) cipher.data,
                                      (uint8_t *) tag.data);
  return out;
}


EverCrypt_Bytes_maybe_plaintext
EverCrypt_Bytes_chacha20_poly1305_decrypt(FStar_Bytes_bytes cipher,
                                          FStar_Bytes_bytes tag,                               
                                          FStar_Bytes_bytes aad,
                                          FStar_Bytes_bytes k,
                                          FStar_Bytes_bytes n) {
  FStar_Bytes_bytes m = {
    .length = cipher.length,
    .data = KRML_HOST_CALLOC(cipher.length, 1)
  };
  uint32_t res =
    EverCrypt_chacha20_poly1305_decrypt((uint8_t *) k.data,
                                        (uint8_t *) n.data,
                                        (uint8_t *) aad.data,
                                        aad.length,
                                        (uint8_t *) m.data,
                                        m.length,
                                        (uint8_t *) cipher.data,
                                        (uint8_t *) tag.data);

  EverCrypt_Bytes_maybe_plaintext out;  
  if (res == 1) {
    out.tag = EverCrypt_Bytes_Correct;
    out._0  = m;
  }
  else {
    out.tag = EverCrypt_Bytes_Error;
  }
  
  return out;
}
